fenicsadapter/adapter_core.py

```diff
@@ -0,0 +1,194 @@
+    """
+
+    filtered_point_sources = dict()
+
+    for point in point_sources.keys():
+        # Filter double boundary points to avoid instabilities and create PointSource
+        if filter_out.inside(point, 1):
+            print("Found a double-boundary point at {location}.".format(location=point))
+        else:
+            filtered_point_sources[point] = point_sources[point]
+
+    return filtered_point_sources
+
+
+def convert_fenics_to_precice(data, sample_points):
+    """Converts FEniCS data of type dolfin.Function into Numpy array for all x and y coordinates on the boundary.
+
+    :param data: FEniCS boundary function
+    :param sample_points: Vertices
+    :raise Exception: if type of data cannot be handled
+    :return: array of FEniCS function values at each point on the boundary
+    """
+    if type(data) is dolfin.Function:
+        x_all, y_all = sample_points[:, 0], sample_points[:, 1]
+        return np.array([data(x, y) for x, y in zip(x_all, y_all)])
+    else:
+        raise Exception("Cannot handle data type %s" % type(data))
+
+
+def get_coupling_boundary_vertices(mesh_fenics, coupling_subdomain, fenics_dimensions, dimensions):
+    """Extracts vertices which lie on the boundary.
+    :return: stack of vertices
+    """
+    n = 0
+    fenics_vertices = []
+    vertices_x = []
+    vertices_y = []
+    if dimensions == 3:
+        vertices_z = []
+
+    if not issubclass(type(coupling_subdomain), SubDomain):
+        raise Exception("no correct coupling interface defined!")
+
+    for v in dolfin.vertices(mesh_fenics):
+        if coupling_subdomain.inside(v.point(), True):
+            n += 1
+            fenics_vertices.append(v)
+            vertices_x.append(v.x(0))
+            if dimensions == 2:
+                vertices_y.append(v.x(1))
+            elif fenics_dimensions == 2 and dimensions == 3:
+                vertices_y.append(v.x(1))
+                vertices_z.append(0)
+            else:
+                raise Exception("Dimensions do not match!")
+
+    assert (n != 0), "No coupling boundary vertices detected"
+
+    if dimensions == 2:
+        return fenics_vertices, np.stack([vertices_x, vertices_y], axis=1), n
+    elif dimensions == 3:
+        return fenics_vertices, np.stack([vertices_x, vertices_y, vertices_z], axis=1), n
+
+
+def are_connected_by_edge(v1, v2):
+    """Returns true if both vertices are connected by an edge. """
+    for edge1 in dolfin.edges(v1):
+        for edge2 in dolfin.edges(v2):
+            if edge1.index() == edge2.index():  # Vertices are connected by edge
+                return True
+    return False
+
+
+def get_coupling_boundary_edges(mesh_fenics, coupling_subdomain, id_mapping):
+    """Extracts edges of mesh which lie on the boundary.
+    :return: two arrays of vertex IDs. Array 1 consists of first points of all edges
+    and Array 2 consists of second points of all edges
+
+    NOTE: Edge calculation is only relevant in 2D cases.
+    """
```

Comment:
Function parameters are not documented.

Why is it only relevant in 2D? Should this be reflected by the name (`get_coupling_boundary_edges_2D`)? What happens if it is called in 3D?